token-tracker/README.adoc

```diff
@@ -0,0 +1,95 @@
+# KEEP Token Tracker
+
+This is a tool for tracking KEEP Token ownership. It will walk over predefined
+sources of truth and inspect KEEP Token ownership at a specific block number.
+
+The tool returns a map of addresses with their total KEEP Token holdings.
+
+## Sources Of Truth
+
+Holders of the tokens can be externally owned accounts or contracts. It is
+necessary to resolve actual owner for tokens held by a contract.
+
+Following sources are currently defined:
+
+- <<KEEP Token>> contract
+- <<Token Staking>> contract
+- <<Token Grant>> contract
+- <<Liquidity Rewards>> contracts
+
+The list of supported sources of truth can be enhanced. For guidelines on 
+implementing additional sources please see <<Development>> section.
+
+
+### KEEP Token
+
+Scans `KeepToken` contract for any historic accounts that have held KEEP token.
+Iterates over the set of addresses and checks KEEP token balance at the specific
+target block.
+
+To avoid duplication with results of other source of truths lookups it ignores
+addresses that are contracts handled by the other sources. For the ignored contracts
+separate sources of truth should be implemented.
+
+### Token Staking
+
+Scans `OldTokenStaking` and `TokenStaking` contracts for any historic stakes to
+determine amount of tokens locked in them.
+
+To avoid duplication with results of other source of truths lookups it ignores
+addresses that are contracts handled by the other sources.
+
+### Token Grant
+
+Scans `TokenGrant` contract for grant holders and their balances.
+
+## Run
+
+### Prerequisites
+
+1. It requires node version greater than `14`: `nvm use 14`.
+
+2. Install dependencies: `npm install`.
+
+### Execute
+
+Following environment variables are expected to be configured:
+
+- `ETH_HOSTNAME` (required) - URL for Ethereum API
+
+Following arguments are expected by the program:
+
+- `target-block` (required) - block at which calculations should be made
+
+Command example:
+
+```sh
+ETH_HOSTNAME="wss://eth-mainnet.ws.alchemyapi.io/v2/....." \
+  ETH_ACCOUNT_PRIVATE_KEY=..... \
+  ./bin/inspect-token-ownership.js --target-block 10000000
+```
+
+## Development
+
+New truth of sources can be added by extending `ITruthSource` class. The new class
+should implement `getTokenHoldingsAtTargetBlock` function that will return
+a map of token holding. The source of truth should be registered with 
+`registerTruthSource` function.
+
+It's important that sources of truth do not duplicate the balances, when
+a holding can occur in few of them. E.g. when an owner staked tokens from a grant
+we could end up with duplications in TokenStaking and TokenGrant. To handle such
+situation we ignore any non externally owned accounts in TokenStaking.
+
+## Known Issues
+
+### error: artifact does not define network *
+
+Web3 provider may return invalid results for `getChainId()` function. The error
+looks like on this example:
```

Comment:
Any idea what is the root cause?

Reply:
Not sure. At first, I had problems when I was using `CacheSubprovider` from `web3-provider-engine`. After I removed it the problem disappeared.... for some time. I had no chance to dig any further.